Adjust makefile

bartekpacia opened the pull request:
Adjusted makefile to match the new directory structure introduced in #21 

main changes:
- renamed main target to `lidarvis`
- renamed the generated executable name to `lidarvis`

# Status

### "pure" build
- [x] Linux
- [x] macOS
- [?] Windows (probably, needs to be checked)

### with `USING_SFML` builds
- [x] Linux
- [ ] macOS - executable is generated, but it crashes immediately after running with the following error message:
```
dyld: Library not loaded: @rpath/libsfml-graphics.2.5.dylib
  Referenced from: /Users/bartek/dev/cpp/lidar-visualizations/./lidarvis
  Reason: image not found
[1]    4982 abort      ./lidarvis
```
- [ ] Windows (probably, needs to be checked)

### with `USING_RPLIDAR`
- [ ] Linux - g++ doesn't compile rplidar_sdk because it finds errors in the library's source code (`rplidar_sdk/sdk/sdk/include/rplidar_driver.h` file).
    To see those errors, `make` in `rplidar_sdk/sdk`. This is very strange because it __was__ working.
- [x] macOS
- [ ] Windows (probably, needs to be checked)

bartekpacia committed:
adjust Makefile to recent changes

bartekpacia committed:
successfully compile with RPLIDAR_SDK (moved hal to rplidar include)

bartekpacia committed:
improve rplidar_sdk install script

bartekpacia committed:
improve sfml install script

bartekpacia committed:
update exec name

bartekpacia committed:
get rid of rplidar_sdk deps, once again

bartekpacia committed:
don't do unnecessary rebuilding

bartekpacia committed:
add a simple Linux installation guide to README

bartekpacia commented:
@dsonyy please
- check whether it's possible on Linux to compile `rplidar_sdk/sdk` using `make` 
- check whether windows builds...build

We also need to find a simple way to conditionally add `USING_SFML` and `USING_RPLIDAR` from command line (currently it's done through modifying the Makefile)

I'm done for today, I'm starting to hate this

dsonyy commented:
## `install_sfml`
- Ubuntu: **OK**

## `install_rplidar`
- Ubuntu: **OK**

## "pure" build
- Windows (make): **OK**
- Ubuntu: **OK**

## with `USING_SFML` build
- Windows (make): **undefined references**
- Ubuntu: **OK**

## with `USING_RPLIDAR` build
- Windows (make):  **undefined references**
- Ubuntu: **OK** but:
  - `Darwin` -> `Linux` (RPLIDAR SDK lib path)
  - added `-pthread` (required by RPLIDAR SDK),
  - you forgot about `-lrplidar_sdk`,
  - RPLIDAR SDK requires to include `/sdk/sdk/include` and `/sdk/sdk/src` (that's weird but official examples do it). I've changed a few Makefile lines a bit from `-I$(RPLIDAR)/sdk/sdk/src/hal` to `-I$(RPLIDAR)/sdk/sdk/src`.

final Makefile (ugly but works):
```
...
lidarvis: main.o app.o cloud.o cloud-grabbers.o cloud-writers.o guis.o scenarios.o
	$(CXX) --output lidarvis \
	main.o \
	app.o \
	cloud.o \
	cloud-grabbers.o \
	cloud-writers.o \
	guis.o \
	scenarios.o \
	-pthread \
	-L$(SFML)/lib \
	-L$(RPLIDAR)/sdk/output/Linux/Release \
	-lrplidar_sdk \
	$(LIBS)

main.o: src/main.cpp
	$(CXX) $(CPPFLAGS) -c src/main.cpp -I$(SFML)/include -I$(RPLIDAR)/sdk/sdk/include -I$(RPLIDAR)/sdk/sdk/src

app.o: src/app.cpp
	$(CXX) $(CPPFLAGS) -c src/app.cpp -I$(SFML)/include -I$(RPLIDAR)/sdk/sdk/include -I$(RPLIDAR)/sdk/sdk/src

cloud.o: src/cloud.cpp
	$(CXX) $(CPPFLAGS) -c src/cloud.cpp

cloud-grabbers.o: src/cloud-grabbers.cpp
	$(CXX) $(CPPFLAGS) -c src/cloud-grabbers.cpp -I$(RPLIDAR)/sdk/sdk/include -I$(RPLIDAR)/sdk/sdk/src
....
```

## with `USING_SFML` and `USING_RPLIDAR` build
- Ubuntu: **OK** but after changes described earlier.

## VS Project
- Windows: **OK**

bartekpacia committed:
Merge branch 'main' into adjust_makefile

bartekpacia committed:
add dsonyy's suggestions

in case rplidar_sdk compilation fails with -Wnarrowing:
add "-Wno-narrowing" to "CXXFLAGS" on line 57 in
rplidar_sdk/sdk/mak_def.inc

bartekpacia commented:
done & working

final status available internally [here](https://www.notion.so/Zbudowa-projekt-na-nix-z-u-yciem-make-cmake-lidar-visualizations-400ed2fd8214469ca921e47dfb29527b)
